### ISSUE TITLE
[red-knot] Support `builtins.callable`

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ruff#17530]
Add support for narrowing on `builtins.callable`, requires `TypeIs` support.

`builtins.pyi`:
```py
def callable(obj: object, /) -> TypeIs[Callable[..., object]]: ...
```

Otherwise, it causes a few false positives, like:
```py
def _(a: Any | None) -> None:
    # /private/tmp/mypy_primer/projects/bidict/bidict/_iter.py:50:34: Object of type `None` is not callable
    if callable(a):
        a()  # red-knot: Object of type `None` is not callable [lint:call-non-callable]
        reveal_type(a)  # revealed_type: Any | None
```

**Comment by AlexWaygood** [comment]
Hmm, won't this fall out naturally from TypeIs support? I'm not sure we'll need to add any special handling for `builtins.callable()` once we understand `TypeIs`.

It doesn't look like we have a dedicated issue for `TypeGuard` and `TypeIs` support right now, but we should maybe make one. It was one of the items listed in https://github.com/astral-sh/ruff/issues/13694, but that issue is now closed

**Comment by dhruvmanila** [comment]
> Hmm, won't this fall out naturally from TypeIs support? I'm not sure we'll need to add any special handling for `builtins.callable()` once we understand `TypeIs`.

Yeah, you're right. There's https://github.com/astral-sh/ruff/pull/16314 which is in draft.

**Comment by carljm** [comment]
`TypeIs` at least should not be too difficult to build on top of our existing narrowing support; tagging "help wanted" for that.

`TypeGuard` may be somewhat more challenging, since it can fully override the previous type (like a cast), meaning we will have to extend our narrowing to support forms of narrowing that don't take the form of intersections. (Probably by changing narrowing constraints from a type to an enum where one variant is "a type to intersect with", another variant is "a typeguard type to override with", etc.)

**Comment by carljm** [comment]
Ah, I wasn't even aware of that PR; I don't get notifications on draft PRs unless pinged. It looks like point 3 there is the same one I mentioned above: `TypeGuard` will require some significant additions to the current implementation of narrowing in red-knot. The differences in implementation are major enough that I would definitely suggest tackling them in separate PRs. I will try to find some time soon to review that PR and see if I can understand the other two points mentioned.

**Comment by carljm** [comment]
> As a note to myself, here's a small edge case example to add to the tests

FWIW, neither pyright nor mypy support this case, and I don't think we need to either. It isn't clear in the spec, but I don't see any evidence in the spec or in the other typechecker implementations to suggest that type guards should have any effect on expressions outside the scope where the type guard is called.
